(DOCSP-26995) Reframe RN SDK open and close realm page (#2529)
mongodb/docs-realm: source/sdk/react-native/realm-database/configure-a-realm.txt

.. _react-native-configure-realm:

=======================================
Configure a Realm - React Native SDK
=======================================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

The ``@realm/react`` library exposes realms in your application using 

> `React Context objects <https://reactjs.org/docs/context.html>`__ and Provider 
> components. You can access realms with React hooks.

Reviewer: ```suggestion
`React Context objects and Provider components <https://reactjs.org/docs/context.html>`__. You can access realms with React hooks.
```

think this makes sense since both part of react?

Author: Ah, yeah, I like that!

To configure and open a local realm, create a Context with the realm's configuration. The 
Context exports a ``RealmProvider`` component that exposes a configured realm. 
All child components of ``RealmProvider`` can access the realm using hooks.

To learn how to open a realm using Device Sync, refer to 
:ref:`Open a Synced Realm <react-native-open-a-synced-realm>`.

> Create a Context Object
> -----------------------

Reviewer: to be more task oriented, how about: 

```suggestion
Create Realm Configuration
--------------------------
```

By passing a :js-sdk:`Configuration <Realm.html#~Configuration>` object to 
``createRealmContext()``, you create a Context object that contains an open 
realm and some hooks that give you access to the realm.

The following example creates and exports a Context object so that it can be 
used elsewhere.

.. literalinclude:: /examples/generated/react-native/js/RealmConfig.snippet.create-realm-context.js
   :language: javascript
   :caption: RealmConfig.js

> Open a Synced Realm
> ~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
Configure a Synced Realm
~~~~~~~~~~~~~~~~~~~~~~~~
```

> Configure a Realm with ``<RealmProvider>``

Reviewer: ```suggestion
Expose a Realm with ``<RealmProvider>``
```

not sure if 'configure' is the right word here since you're not actually making and config here. maybe 'expose' better?

To open a realm that synchronizes data with Atlas using Device Sync,
refer to :ref:`Open a Synced Realm <react-native-open-a-synced-realm>`.

> Open an In-Memory Realm
> ~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
Configure an In-Memory Realm
~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

.. literalinclude:: /examples/generated/react-native/ts/configure-realm-local.test.snippet.configure-realm.tsx
   :language: javascript

Open a Synced Realm
~~~~~~~~~~~~~~~~~~~

To open a realm that synchronizes data with Atlas using Device Sync,
refer to :ref:`Open a Synced Realm <react-native-open-a-synced-realm>`.

Open an In-Memory Realm
~~~~~~~~~~~~~~~~~~~~~~~

To create a realm that runs entirely in memory without being written to a file,
add ``inMemory: true`` to your :js-sdk:`Realm.Configuration
<Realm.html#~Configuration>` object:

.. literalinclude:: /examples/generated/react-native/ts/index.snippet.in-memory-realm.ts
   :language: javascript
   :emphasize-lines: 3

In-memory realms may use disk space if memory is running low, but files created 

> by an in-memory realm are deleted when you close the realm.

Reviewer: rather than these 2 sections being under 'Configure a Realm with `RealmProvider`', i think it'd make more sense for these to be subsections of the 'Create a Context Object with createRealmContext' section since they're just about creating that context

.. _react-native-configure-realm:

=======================================
Configure a Realm - React Native SDK

> =======================================

Reviewer: ah i think my question is answered here

https://github.com/mongodb/docs-realm/pull/2529/files#diff-e2ba0caf7cac0d6a837742cca53c7a81a73941d6c5f1ebbc32e9402263920e70R16-R17

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

> The ``@realm/react`` library uses `React Context objects <https://reactjs.org/docs/context.html>`__

Reviewer: ```suggestion
The ``@realm/react`` library uses `React Context <https://reactjs.org/docs/context.html>`__
```

to create and configure realms. You can access realms with React hooks.

``@realm/react`` automatically opens and closes realms for you. To manage how it
does so, you need to pass props to the ``@realm/react`` Context Providers:

- ``<AppProvider>``
- ``<UserProvider>``

> - ``<RealmProvider>``

Reviewer: rather then mentioning all these context providers, could it make sense to just mention `RealmProvider`? 

the other 2 are only needed for realm w sync.

For Flexible Sync realms, you need all three Providers, nested in the order above.

> For local-only realms, you only need ``<RealmProvider>``.

Reviewer: can you refactor this content to say smthn like: 

- to open local realm, only need `RealmProvider`
- to use realm with atlas device sync, need to wrap the `RealmProvider` in `AppProvider` and `UserProvider`. to learn more about using Device Sync with @realm/react, refer to ...<some link>

i think it'd help clarify the page to make it local-first, and make clear sync is optional and kinda different process.

> The ``@realm/react`` library uses `React Context objects <https://reactjs.org/docs/context.html>`__
> to create and configure realms. You can access realms with React hooks.
> 
> ``@realm/react`` automatically opens and closes realms for you. To manage how it
> does so, you need to pass props to the ``@realm/react`` Context Providers:
> 
> - ``<AppProvider>``
> - ``<UserProvider>``
> - ``<RealmProvider>``
> 
> For Flexible Sync realms, you need all three Providers, nested in the order above.
> For local-only realms, you only need ``<RealmProvider>``.

Reviewer: i think we can consolidate these paragraphs a fair bit. first, i think this page and the intro for it should focus much more on only local realms. and with that focus, i think the flow of the info could be more streamlined to something like:

```suggestion
The ``@realm/react`` library exposes realms throughout your application with `React Context objects <https://reactjs.org/docs/context.html>`__ and Provider components. You can access realms with React hooks.

To open a local realm, create a Context with the realm's configuration. The Context exports a `RealmProvider` component that exposes a realm with the configuration passed to the Context. All child components of the `RealmProvider` can access the realm using hooks.

To learn how to open a realm using Device Sync, refer to :ref:`Open a Synced Realm <TODO: add relevant ref>`.
```

Create a Context Object with ``createRealmContext``
---------------------------------------------------

By passing a :js-sdk:`Configuration <Realm.html#~Configuration>` object to 
``createRealmContext``, you can create a Context object that contains an open 
realm and some hooks that give you access to the realm.

> In this example, we create and export our config object so that it can be used elsewhere.

Reviewer: ```suggestion
In this example, we create and export our Context object so that it can be used elsewhere.
```

.. literalinclude:: /examples/generated/react-native/js/index.snippet.create-realm-context.js

>    :language: javascript

Reviewer: ```suggestion
   :language: javascript
   :caption: Models.jsx
```

i think having something like this can help clarify the following example which imports this file.

Author: Great suggestion! I added this caption:
`:caption: Models/index.js`

Reviewer: rather than `Models/index.js`, why not just `Models.js`?

Author: Updated to `RealmConfig.js` based on the other discussion.

> Configure a realm with ``<RealmProvider>``

Reviewer: ```suggestion
Configure a Realm with ``<RealmProvider>``
```

i think uppercase here b/c in a title 🤔

------------------------------------------

> ``<RealmProvider>`` is a wrapper for your Realm Database. All child components

Reviewer: ```suggestion
``<RealmProvider>`` is a wrapper that exposes a Context with Realm Database to its child components. All child components
```

inside ``<RealmProvider>`` have access to hooks that let you read, write, and update
data. This is what a local-only realm wrapper might look like:

.. literalinclude:: /examples/generated/react-native/js/configure-realm-local.test.snippet.configure-realm.jsx
   :language: javascript

Enable Flexible Sync
~~~~~~~~~~~~~~~~~~~~

To use a Flexible Sync realm, you must nest ``<RealmProvider>`` within ``<AppProvider>`` 
and ``<UserProvider>``. This ensures your client has access to your App Services App
and that a user with permissions exists to access your realm data.

.. literalinclude:: /examples/generated/react-native/js/configure-realm-sync.test.snippet.configure-realm-sync.jsx

>    :language: javascript

Reviewer: i think this section should probably be removed. to really get going w device sync, you need some more context, like app creation on the backend, enable flexible sync, etc. 

this probably needs to be a different page. see [flutter](https://www.mongodb.com/docs/realm/sdk/flutter/sync/open-synced-realm/#std-label-flutter-open-synced-realm) and [swift](https://www.mongodb.com/docs/realm/sdk/swift/sync/configure-and-open-a-synced-realm/) sdks for example.

while it's def important to mention opening a synced realm on this page, i think it'd suffice to point to the 'open a synced realm' on this page. looks like there's already a ticket for @realm/reactifing that page, [DOCSP-27009](https://jira.mongodb.org/browse/DOCSP-27009)/). 

to borrow some text for this section, you can [see what's on the flutter page](https://www.mongodb.com/docs/realm/sdk/flutter/realm-database/configure-and-open/#open-a-synced-realm).

> 

Reviewer: also, what about in memory realms? should that be on this page?

Author: Yeah, I'm looking into adding that back to this page.

Author: I've added this section back in.

> Open more than one realm

Author: Agh. This title casing on headings. I'll get it eventually. Old habits die hard.

I think I'm going to go with "Open One or More Realms" though. Because you might want to open just two. And two is not multiple.

> Open more than one realm

Reviewer: in that case, what about 'More than One' like you originally had it? 

since you already talk about just the 'one realm' case above

~~~~~~~~~~~~~~~~~~~~~~~~

> You can open more than one realm at a time by creating another Context object.

Reviewer: ```suggestion
You can open more than one realm at a time by creating additional Contexts and ``RealmProvider`` components.
```

.. literalinclude:: /examples/generated/react-native/js/configure-realm-sync.test.snippet.two-realm-contexts.jsx
   :language: javascript

Key Concept: Realm Files

> ------------------------

Reviewer: agree w you that this should be refactored and that it doesn't make sense to do as part of this PR
